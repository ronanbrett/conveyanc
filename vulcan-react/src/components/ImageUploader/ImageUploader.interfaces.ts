export interface ImageType {
  dataURL?: string;
  file?: File;
  [key: string]: any;
}

export type ImageListType = Array<ImageType>;

export interface AdditionErrorProps {
  fileList: any;
  keyUpdate?: any;
}

export interface ImageUploadingPropsType {
  value: ImageListType;
  onChange: (value: ImageListType, addUpdatedIndex?: Array<number>) => void;
  children?: (props: ExportInterface) => React.ReactNode;
  multiple?: boolean;
  maxNumber?: number;
  acceptType?: Array<string>;
  maxFileSize?: number;
  resolutionWidth?: number;
  resolutionHeight?: number;
  resolutionType?: ResolutionType;
  onError?: (errors: ErrorsType, files?: ImageListType) => void;
  dataURLKey?: string;
  [param: string]: any;
}

export interface ExportInterface {
  imageList: ImageListType;
  onImageUpload: () => void;
  onImageRemoveAll: () => void;
  errors: ErrorsType;
  onImageUpdate: (index: number) => void;
  onImageRemove: (index: number) => void;
  isDragging: Boolean;
  dragProps: {
    onDrop: (e: any) => void;
    onDragEnter: (e: any) => void;
    onDragLeave: (e: any) => void;
    onDragOver: (e: any) => void;
  };
}

export type ErrorsType = {
  maxFileSize?: boolean;
  maxNumber?: boolean;
  acceptType?: boolean;
  resolution?: boolean;
} | null;

export type ResolutionType = "absolute" | "less" | "more" | "ratio";
