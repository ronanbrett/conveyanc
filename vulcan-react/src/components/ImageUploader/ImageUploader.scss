@import "src/styles/vars";

.ImageUploader {
  &__drop {
    width: 100%;
    height: 200px;
    background: transparent;
    border: 2px dashed var(--border-color);
    margin-bottom: var(--spacing-m);
  }

  &__images {
    display: grid;
    grid-template-columns: 1fr 1fr 1fr 1fr;
    grid-template-rows: auto;
  }

  &__image {
    position: relative;

    img {
      width: 100%;
      display: block;
      object-fit: cover;
      height: 100%;
    }
  }

  &--PENDING {
    opacity: 0.4;
  }

  &--FAILED {
    border: 2px solid red;
    opacity: 0.4;
  }

  &__image-actions {
    position: absolute;
    top: 0;
    left: 0;
    color: white;

    button {
      color: white;
    }
  }
}
