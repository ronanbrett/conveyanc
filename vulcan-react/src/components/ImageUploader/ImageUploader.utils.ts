import { MutableRefObject } from "react";
import { v4 } from "uuid";

export interface ImageType {
  dataURL?: string;
  file?: File;
  [key: string]: any;
}

export type ImageListType = Array<ImageType>;

export interface ImageUploadingPropsType {
  value: ImageListType;
  onChange: (value: ImageListType, addUpdatedIndex?: Array<number>) => void;
  children?: (props: ExportInterface) => React.ReactNode;
  multiple?: boolean;
  maxNumber?: number;
  acceptType?: Array<string>;
  maxFileSize?: number;
  resolutionWidth?: number;
  resolutionHeight?: number;
  resolutionType?: ResolutionType;
  onError?: (errors: ErrorsType, files?: ImageListType) => void;
  dataURLKey?: string;
}

export interface ExportInterface {
  imageList: ImageListType;
  onImageUpload: () => void;
  onImageRemoveAll: () => void;
  errors: ErrorsType;
  onImageUpdate: (index: number) => void;
  onImageRemove: (index: number) => void;
  isDragging: Boolean;
  dragProps: {
    onDrop: (e: any) => void;
    onDragEnter: (e: any) => void;
    onDragLeave: (e: any) => void;
    onDragOver: (e: any) => void;
  };
}

export type ErrorsType = {
  maxFileSize?: boolean;
  maxNumber?: boolean;
  acceptType?: boolean;
  resolution?: boolean;
} | null;

export type ResolutionType = "absolute" | "less" | "more" | "ratio";
export const openFileDialog = (
  inputRef: MutableRefObject<HTMLElement>
): void => {
  if (inputRef.current) inputRef.current.click();
};

export const getAcceptTypeString = (acceptType?: Array<string>) => {
  return acceptType && acceptType.length > 0
    ? acceptType.map((item) => `.${item}`).join(", ")
    : "image/*";
};

export const getBase64 = (file: File): Promise<string> => {
  const reader = new FileReader();
  return new Promise((resolve) => {
    reader.addEventListener("load", () => resolve(String(reader.result)));
    reader.readAsDataURL(file);
  });
};

export const getListFiles = (
  files: FileList,
  dataURLKey: string
): Promise<ImageListType> => {
  const promiseFiles: Array<Promise<string>> = [];
  for (let i = 0; i < files.length; i++) {
    promiseFiles.push(getBase64(files[i]));
  }
  return Promise.all(promiseFiles).then((fileListBase64: Array<string>) => {
    const fileList: ImageListType = fileListBase64.map((base64, index) => ({
      id: v4(),
      pending: true,
      complete: false,
      [dataURLKey]: base64,
      file: files[index],
    }));
    return fileList;
  });
};
