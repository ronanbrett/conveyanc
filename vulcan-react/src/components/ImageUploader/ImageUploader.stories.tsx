import React from 'react';

import ImageUploader from './ImageUploader';

import { action } from '@storybook/addon-actions';

export default {
  title: 'Components/ImageUploader',
  decorators: [],
  component: ImageUploader,
};


export const Default = (props) => (
  <ImageUploader {...props}></ImageUploader>
);
