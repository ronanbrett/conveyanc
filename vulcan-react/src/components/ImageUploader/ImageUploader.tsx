import { S3Level, S3ObjectArgs } from "@core/api/graphql";
import { Storage } from "@services/aws.service";
import classNames from "classnames";
import { IconButton } from "@components";
import { FieldConfig, useField } from "formik";
import { reduce } from "lodash-es";
import React, { FC, useEffect, useState } from "react";
import { empty, forkJoin, from, Subject } from "rxjs";
import { catchError, map, scan, switchMap } from "rxjs/operators";
import ImageUploading from "./ImageUploader.base";
import "./ImageUploader.scss";

interface ImageI extends Partial<S3ObjectArgs> {
  data_url?: string;
  file?: File;
  id?: string;
  index?: number;
  failed?: boolean;
  pending?: boolean;
  complete?: boolean;
}

interface ImageUploaderProps {
  children?: any;
}

const ImageUploader: FC<ImageUploaderProps & FieldConfig> = ({ ...props }) => {
  const [field, meta, helpers] = useField(props);

  const [images, setImages] = React.useState([]);
  const [imageUploadMap, setImageUploadMap] = useState<any>({});

  const [uploadQueue$] = useState(() => {
    // Arrow function is used to init Singleton Subject. (in a scope of a current component)
    return new Subject<any>();
  });

  useEffect(() => {
    if (field.value && field.value.length) {
      const vals = reduce(
        field.value,
        (accum: any, field: any) => {
          accum[field.key] = {
            ...field,
            complete: true,
            id: field.key,
          };
          return accum;
        },
        {}
      );

      const images = field.value.map((field: any) => ({
        ...field,
        id: field.key,
      }));
      setImageUploadMap(vals);
      setImages(images);
    }
  }, []);

  useEffect(() => {
    const completedImages = reduce(
      images,
      (accum: ImageI[], image: ImageI) => {
        const mappedImg = imageUploadMap[image.id];
        if (mappedImg && mappedImg.complete) {
          const { key, level, directory, bucket, region } = mappedImg;
          const img = {
            key,
            level,
            directory,
            bucket,
            region,
            data_url: image.data_url,
          };
          accum.push(img);
        }
        return accum;
      },
      []
    );

    helpers.setValue(completedImages);
  }, [images, imageUploadMap]);

  useEffect(() => {
    const DIRECTORY = "property";
    const subscription = uploadQueue$
      .pipe(
        switchMap((images: ImageI[]) =>
          forkJoin(
            images.map((image) =>
              from(
                Storage.put(`${DIRECTORY}/${image.id}`, image.file, {
                  level: "public",
                }).catch((err) => {
                  image.failed = true;
                  return image;
                })
              ).pipe(
                map((res) => ({
                  ...res,
                  id: image.id,
                  key: `${DIRECTORY}/${image.id}`,
                  level: S3Level.Public,
                  region: CONFIG.AWS.REGION,
                  bucket: CONFIG.AWS.S3.BUCKET,
                  pending: false,
                  complete: true,
                })),
                catchError((err, obs) => {
                  return empty();
                })
              )
            )
          )
        ),
        scan((accum: any, images: ImageI[]) => {
          const statusMap = reduce(
            images,
            (accum: any, image, index) => {
              accum[image.id] = image;
              return accum;
            },
            {}
          );

          accum = { ...accum, ...statusMap };

          return accum;
        }, {})
      )
      .subscribe((images: ImageI[]) => {
        setImageUploadMap(images);
      });

    return () => {
      // On Component destroy. notify takeUntil to unsubscribe from current running ajax request
      // unsubscribe filter change listener
      subscription.unsubscribe();
    };
  }, []);

  const maxNumber = 16;
  const onChange = (imageList: ImageI[], addUpdateIndex?: number[]) => {
    // data for submit
    setImages(imageList);

    if (addUpdateIndex?.length) {
      uploadQueue$.next(
        addUpdateIndex.map((index) => ({ ...imageList[index], index }))
      );
    }
  };

  return (
    <div className="ImageUploader">
      <ImageUploading
        multiple
        value={images}
        onChange={onChange}
        maxNumber={maxNumber}
        dataURLKey="data_url"
      >
        {({
          imageList,
          onImageUpload,
          onImageRemoveAll,
          onImageUpdate,
          onImageRemove,
          isDragging,
          dragProps,
        }) => (
          // write your building UI
          <div className="ImageUploader__wrapper">
            <button
              className="ImageUploader__drop"
              type="button"
              style={isDragging ? { color: "red" } : null}
              onClick={onImageUpload}
              {...dragProps}
            >
              Click or Drop here
            </button>

            <div className="ImageUploader__images">
              {imageList.map((image: any, index) => {
                const state = imageUploadMap[image.id] ?? { pending: true };

                const cx = classNames({
                  ImageUploader__image: true,
                  "ImageUploader--FAILED": state.failed,
                  "ImageUploader--PENDING": state.pending,
                  "ImageUploader--complete": state.complete,
                });
                return (
                  <div key={index} className={cx}>
                    <img src={image.data_url} alt="" />
                    <div className="ImageUploader__image-actions">
                      <IconButton
                        size="mini"
                        icon="delete"
                        onClick={() => onImageRemove(index)}
                      ></IconButton>
                    </div>
                  </div>
                );
              })}
            </div>
          </div>
        )}
      </ImageUploading>
    </div>
  );
};

export default ImageUploader;
