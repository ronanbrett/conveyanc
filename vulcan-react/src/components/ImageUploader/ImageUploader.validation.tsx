import { DEFAULT_NULL_INDEX } from "./ImageUploader.consts";
import {
  ResolutionType,
  ErrorsType,
  ImageUploadingPropsType,
  AdditionErrorProps,
} from "./ImageUploader.interfaces";

export const isResolutionValid = (
  dataURL: string,
  resolutionType: ResolutionType,
  resolutionWidth: number = 0,
  resolutionHeight: number = 1
): Promise<boolean> => {
  const image = new Image();

  return new Promise((resolve) => {
    image.addEventListener("load", () => {
      if (image.width && image.height) {
        switch (resolutionType) {
          case "absolute": {
            if (
              image.width === resolutionWidth &&
              image.height === resolutionHeight
            )
              return resolve(true);
            return resolve(false);
          }
          case "ratio": {
            const ratio = resolutionWidth / resolutionHeight;
            if (image.width / image.height === ratio) return resolve(true);
            return resolve(false);
          }
          case "less": {
            if (
              image.width <= resolutionWidth &&
              image.height <= resolutionHeight
            )
              return resolve(true);
            return resolve(false);
          }
          case "more": {
            if (
              image.width >= resolutionWidth &&
              image.height >= resolutionHeight
            )
              return resolve(true);
            return resolve(false);
          }
        }
      }
      resolve(true);
    });
    image.src = dataURL;
  });
};

export const isImageValid = (fileType: string) => {
  if (fileType.includes("image")) {
    return true;
  }
  return false;
};

export const isMaxFileSizeValid = (fileSize: number, maxFileSize: number) => {
  return fileSize <= maxFileSize;
};

export const isAcceptTypeValid = (acceptType: string[], fileName: File) => {
  if (acceptType && acceptType.length > 0) {
    const type: string = fileName.name.split(".").pop() || "";
    if (
      acceptType.findIndex(
        (item) => item.toLowerCase() === type.toLowerCase()
      ) < 0
    ) {
      return false;
    }
  }
  return true;
};

export const isMaxNumberValid = (
  totalNumber: number,
  maxNumber: number,
  keyUpdate: any
) => {
  if (keyUpdate === DEFAULT_NULL_INDEX && maxNumber && totalNumber > maxNumber)
    return false;
  return true;
};

export const getErrorValidation = async ({
  fileList,
  keyUpdate,
  value,
  maxNumber,
  dataURLKey,
  acceptType,
  maxFileSize,
  resolutionType,
  resolutionWidth,
  resolutionHeight,
}: Partial<ImageUploadingPropsType & AdditionErrorProps>): Promise<
  ErrorsType
> => {
  const newErrors: ErrorsType = {};
  if (!isMaxNumberValid(fileList.length + value.length, maxNumber, keyUpdate)) {
    newErrors.maxNumber = true;
  } else {
    for (let i = 0; i < fileList.length; i++) {
      const { file, [dataURLKey]: dataURL } = fileList[i];
      if (file) {
        if (!isImageValid(file.type)) {
          newErrors.acceptType = true;
          break;
        }
        if (!isAcceptTypeValid(acceptType, file.name)) {
          newErrors.acceptType = true;
          break;
        }
        if (!isMaxFileSizeValid(file.size, maxFileSize)) {
          newErrors.maxFileSize = true;
          break;
        }
      }
      if (dataURL && resolutionType) {
        const checkRes = await isResolutionValid(
          dataURL,
          resolutionType,
          resolutionWidth,
          resolutionHeight
        );
        if (!checkRes) {
          newErrors.resolution = true;
          break;
        }
      }
    }
  }
  if (Object.values(newErrors).find(Boolean)) return newErrors;
  return null;
};
