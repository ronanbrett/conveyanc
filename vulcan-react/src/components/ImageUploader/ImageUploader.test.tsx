import React from "react";
import { create } from "react-test-renderer";
import ImageUploader from "./ImageUploader";
import { Formik, Form } from "formik";

describe("ImageUploader", () => {
  it("renders without error", () => {
    const component = create(
      <Formik initialValues={{}} onSubmit={(x) => console.log(x)}>
        <Form>
          <label htmlFor="propertyType">Property Type</label>
          <ImageUploader name="images"></ImageUploader>
        </Form>
      </Formik>
    );

    expect(component).toMatchSnapshot();
  });
});
