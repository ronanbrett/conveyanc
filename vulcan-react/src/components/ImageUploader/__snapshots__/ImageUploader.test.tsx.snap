// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageUploader renders without error 1`] = `
<form
  action="#"
  onReset={[Function]}
  onSubmit={[Function]}
>
  <label
    htmlFor="propertyType"
  >
    Property Type
  </label>
  <div
    className="ImageUploader"
  >
    <input
      accept="image/*"
      multiple={true}
      onChange={[Function]}
      style={
        Object {
          "display": "none",
        }
      }
      type="file"
    />
    <div
      className="ImageUploader__wrapper"
    >
      <button
        className="ImageUploader__drop"
        onClick={[Function]}
        onDragEnter={[Function]}
        onDragLeave={[Function]}
        onDragOver={[Function]}
        onDrop={[Function]}
        style={null}
        type="button"
      >
        Click or Drop here
      </button>
      <div
        className="ImageUploader__images"
      />
    </div>
  </div>
</form>
`;
